packages/components/src/FilterDateRangePicker/subcomponents/FilterDateRangePickerField/FilterDateRangePickerField.tsx

```diff
@@ -120,74 +105,140 @@ export const FilterDateRangePickerField = ({
+    inputEndValue: transformDateToInputValue(selectedRange?.to) || "",
+    startMonth:
+      selectedRange?.from && !isInvalidDate(selectedRange?.from)
+        ? selectedRange?.from
+        : defaultMonth || new Date(),
+  })
+
   const inputStartDateHandlers = useDateInputHandlers({
     locale,
     disabledDays,
-    setInputValue: setInputStartDateValue,
+    setInputValue: value => {
+      dispatch({
+        type: "update_input_start_field",
+        inputValue: value as string,
+      })
+    },
     onDateChange: date => {
-      const newDate = validateStartDate(date)
-      const endDate = parseDateFromTextFormatValue(inputEndDateValue, locale)
+      const newDate = validateStartDate(date, state.inputStartValue)
+      const endDate = parseDateFromTextFormatValue(state.inputEndValue, locale)
+
+      dispatch({
+        type: "update_selected_start_date",
+        date: newDate,
+      })
+
+      dispatch({
+        type: "update_selected_end_date",
+        date: endDate,
+      })
 
       if (newDate && !isInvalidDate(endDate)) {
         const newEndDate = dateEndValidation.validateEndDateBeforeStartDate({
           startDate: newDate,
           startDateFieldLabel: inputStartDateLabel,
           endDate,
-          endDateInputValue: inputEndDateValue,
+          endDateInputValue: state.inputEndValue,
         })
         handleDateRangeChange({ from: newDate, to: newEndDate })
       } else {
-        handleDateRangeChange({ from: newDate, to: rangeEnd })
+        handleDateRangeChange({ from: newDate, to: state.selectedEndDate })
       }
-
-      if (newDate) setStartMonth(newDate)
     },
     ...inputStartDateProps,
   })
 
   const inputEndDateHandlers = useDateInputHandlers({
     locale,
     disabledDays,
-    setInputValue: setInputEndDateValue,
+    setInputValue: value => {
+      dispatch({
+        type: "update_input_end_field",
+        inputValue: value as string,
+      })
+    },
     onDateChange: date => {
-      const newDate = validateEndDate(date)
-      handleDateRangeChange({ from: rangeStart, to: newDate })
+      const newDate = validateEndDate(date, state.inputEndValue)
+
+      dispatch({
+        type: "update_selected_end_date",
+        date: newDate,
+      })
+
+      handleDateRangeChange({ from: state.selectedStartDate, to: newDate })
     },
     ...inputEndDateProps,
   })
 
   const handleCalendarSelectRange: CalendarRangeProps["onSelect"] = range => {
-    const newStartDate = validateStartDate(range?.from)
-    const newEndDate = validateEndDate(range?.to)
-    setInputStartDateValue(transformDateToInputValue(newStartDate))
-    setInputEndDateValue(transformDateToInputValue(newEndDate))
+    const inputStartValue = transformDateToInputValue(range?.from)
+    const inputEndValue = transformDateToInputValue(range?.to)
+
+    const newStartDate = validateStartDate(range?.from, inputStartValue)
+    const newEndDate = validateEndDate(range?.to, inputEndValue)
```

Comment:
I'm pretty sure this was all that really needed to be changed in order for the validation to get the correct input values. But hey, reducers look nicer lol.